### PR TITLE
bug fix and  add new NumericUpDownDemo

### LINKED ISSUE #172: NumericUpDown 如果达到最大值，继续用drag的话。会直接归 0
uint 0xffffffff + 1 = 0

### PR DESCRIPTION
**Pull request by heartacker**
1. _HorizontalContentAlignment
2. new demo 
3. fix #172 
4. should not SyncTextAndValue while fromTextToValue and clamp

### CODE REVIEW

**Comment by heartacker** [comment]
- linux 
image

image

**Comment by rabbitism** [comment]
❤️ will review tomorrow.

File: src/Ursa.Themes.Semi/Controls/NumericUpDown.axaml [hunk 1]
```diff
@@ -35,24 +35,26 @@
...
+                        BorderThickness="{TemplateBinding BorderThickness}"
+                        ShowButtonSpinner="{TemplateBinding ShowButtonSpinner}">
                         <Panel>
                             <TextBox
                                 Name="PART_TextBox"
                                 Height="{TemplateBinding Height}"
                                 MinHeight="{DynamicResource NumericUpDownWrapperDefaultHeight}"
+                                HorizontalContentAlignment="{TemplateBinding HorizontalContentAlignment}"
                                 AcceptsReturn="False"
+                                Background="Transparent"
```
Comment (rabbitism) [inline src/Ursa.Themes.Semi/Controls/NumericUpDown.axaml]: Background should not be transparent.

File: src/Ursa/Controls/NumericUpDown/NumericUpDownBase.cs [hunk 2]
```diff
@@ -47,6 +48,15 @@ public bool IsReadOnly
         set => SetValue(IsReadOnlyProperty, value);
     }
 
+    public static readonly StyledProperty<HorizontalAlignment> HorizontalContentAlignmentProperty = AvaloniaProperty.Register<NumericUpDown, HorizontalAlignment>(
```
Comment (rabbitism) [inline src/Ursa/Controls/NumericUpDown/NumericUpDownBase.cs]: Use `ContentControl.HorizontalContentAlignmentProperty.AddOwner<NumericUpDown>();`

**Review by rabbitism — CHANGES_REQUESTED** [review]

**Review by heartacker — COMMENTED** [review]
> Background should not be transparent.  

我以为 panel 是嵌套到 ButtonSpinner  ，所以以 ButtonSpinner  为主要，我可以改回去 
<ButtonSpinner                       
                        Background="{TemplateBinding Background}"

**Comment by rabbitism** [comment]
> > Background should not be transparent.
> 
> 我以为 panel 是嵌套到 ButtonSpinner ，所以以 ButtonSpinner 为主要，我可以改回去 <ButtonSpinner Background="{TemplateBinding Background}"

所以你是需要一个默认无背景的样式是么……根据Semi Design的理念input都是默认灰色背景的

**Comment by heartacker** [comment]
ready to merge @rabbitism

**Comment by rabbitism** [comment]
Look good to me so far. But I will have one more fix after merging, it's related to the actual behavior of textbox readonly and drag panel visibility.

**Review by rabbitism — APPROVED** [review]

**Pull request merged**